File: sentinel/engine-ts/src/content.ts
/** Content package loading: schema validation, manifest hash verification,
 * semantic cross-checks. Fails closed — any problem raises ContentError with
 * every message collected (prime directives 2 and 4).
 *
 * This module is PURE (no filesystem access): all file reads go through an
 * injected ContentReader, so the core stays dependency-free and portable.
 * Node wiring lives in content_node.ts.
 */
import { canonicalJson, sha256Hex } from "./canonical.js";
import { ENGINE_VERSION, OBSERVATION_TYPES, REFERENCEABLE_FLAGS } from "./constants.js";
import { ContentError } from "./errors.js";
import { validate, type Json } from "./schema_validator.js";

export const TABLE_TYPES = [
  "operational_bounds", "never_ignore", "floors", "baseline_config",
  "population_baselines", "units", "worse_direction", "artifact_rules",
  "global_modifiers", "recheck_intervals", "trajectory_rules", "event_codes",
];

export const COMPARATOR_OPS = ["gte", "lte", "eq"];
export const FUNCTION_OPS = [
  "delta_from_baseline_abs", "delta_from_baseline_pct", "trend_slope",
  "crossed_threshold_count", "sustained_for_min", "event_present",
];

/** Injected file access. All paths are package-relative with forward slashes. */
export interface ContentReader {
  /** Label recorded on the handle (the package path or equivalent). */
  packagePath: string;
  /** UTF-8 text of a package file, or null if it does not exist. */
  readPackageText(rel: string): string | null;
  /** Relative forward-slash paths of ALL files under the package directory. */
  listPackageFiles(): string[];
  /** UTF-8 text of a schema file by base name (e.g. "manifest.schema.json"). */
  readSchemaText(name: string): string;
}

/** Build a ContentReader from an in-memory map of path → file text. */
export function mapReader(
  packagePath: string,
  packageFiles: Record<string, string>,
  schemaFiles: Record<string, string>,
): ContentReader {
  return {
    packagePath,
    readPackageText: (rel: string) => (rel in packageFiles ? packageFiles[rel] : null),
    listPackageFiles: () => Object.keys(packageFiles),
    readSchemaText: (name: string) => {
      if (!(name in schemaFiles)) {
        throw new Error(`schema file missing: ${name}`);
      }
      return schemaFiles[name];
    },
  };
}

export interface ContentHandle {
  package_path: string;
  content_version: string;
  flags: string[];
  tables: Record<string, any>;
  signatures: any[];
  published_signatures: any[];
}

function semverTuple(v: string): number[] {
  return v.split(".").map((x) => {
    const n = parseInt(x, 10);
    if (!/^\d+$/.test(x) || Number.isNaN(n)) {
      throw new Error(`invalid literal for int() with base 10: ${x}`);
    }
    return n;
  });
}

function semverGt(a: number[], b: number[]): boolean {
  // Python tuple comparison: element-wise, shorter tuple is a prefix-loser.
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) {
      return a[i] > b[i];
    }
  }
  return a.length > b.length;
}

/** Flatten a logic tree into its leaf conditions (depth-first, in order). */
export function walkConditions(node: any, out: any[]): void {
  if (typeof node !== "object" || node === null || Array.isArray(node)) {
    return;
  }
  for (const comb of ["all_of", "any_of", "none_of"]) {
    if (comb in node) {
      for (const child of node[comb]) {
        walkConditions(child, out);
      }
      return;
    }
  }
  if ("at_least_n_of" in node) {
    for (const child of node["at_least_n_of"]["of"]) {
      walkConditions(child, out);
    }
    return;
  }
  out.push(node);
}

function checkSignatureSemantics(
  sig: any,
  knownMetrics: Set<string>,
  eventIds: Set<string>,
  enumMetrics: Set<string>,
  errors: string[],
): void {
  const sid = sig.signature_id ?? "?";
  const conditions: any[] = [];
  walkConditions(sig.logic ?? {}, conditions);
  if (conditions.length === 0) {
    errors.push(`${sid}: logic tree has no conditions`);
  }
  const idsSeen = new Set<string>();
  conditions.forEach((cond, i) => {
    const where = `${sid}: condition ${i}`;
    if ("flag" in cond) {
      if (!REFERENCEABLE_FLAGS.includes(cond.flag)) {
        errors.push(`${where}: unknown flag '${cond.flag}'`);
      }
      return;
    }
    if ("baseline_status" in cond && !("op" in cond)) {
      if (!("metric" in cond)) {
        errors.push(`${where}: baseline_status condition requires 'metric'`);
      }
      return;
    }
    const op = cond.op;
    const metric = cond.metric;
    if (op === undefined || op === null) {
      errors.push(`${where}: leaf condition needs 'op', 'flag', or 'baseline_status'`);
      return;
    }
    const cid = cond.id;
    if (cid) {
      if (idsSeen.has(cid)) {
        errors.push(`${sid}: duplicate condition id '${cid}'`);
      }
      idsSeen.add(cid);
    }
    if (op === "event_present") {
      if (!("event_id" in cond)) {
        errors.push(`${where}: event_present requires 'event_id'`);
      } else if (!eventIds.has(cond.event_id)) {
        errors.push(`${where}: event_id '${cond.event_id}' not in controlled vocabulary`);
      }
      if (!("window_min" in cond)) {
        errors.push(`${where}: event_present requires 'window_min'`);
      }
      return;
    }
    if (metric === undefined || metric === null) {
      errors.push(`${where}: op '${op}' requires 'metric'`);
      return;
    }
    if (!knownMetrics.has(metric)) {
      errors.push(`${where}: unknown metric '${metric}'`);
    }
    const comparators = COMPARATOR_OPS.filter((k) => k in cond);
    if (COMPARATOR_OPS.includes(op)) {
      if (!("value" in cond)) {
        errors.push(`${where}: raw-value op '${op}' requires 'value'`);
      }
      if (comparators.length > 0) {
        errors.push(`${where}: raw-value op '${op}' must not also carry comparator keys`);
      }
      if (!("window_min" in cond)) {
        errors.push(`${where}: raw-value op '${op}' requires 'window_min'`);
      }
    } else if (FUNCTION_OPS.includes(op)) {
      if (comparators.length !== 1) {
        errors.push(`${where}: op '${op}' requires exactly one of gte/lte/eq keys`);
      }
      if (!("window_min" in cond)) {
        errors.push(`${where}: op '${op}' requires 'window_min'`);
      }
      if (op === "trend_slope" && !("min_points" in cond)) {
        errors.push(`${where}: trend_slope requires 'min_points'`);
      }
      if (op === "crossed_threshold_count" || op === "sustained_for_min") {
        if (!("threshold" in cond) || !("direction" in cond)) {
          errors.push(`${where}: op '${op}' requires 'threshold' and 'direction'`);
        }
      }
      if (enumMetrics.has(metric) && op !== "event_present" && !COMPARATOR_OPS.includes(op)) {
        errors.push(`${where}: op '${op}' not valid for ordinal metric '${metric}'`);
      }
    } else {
      errors.push(`${where}: unknown op '${op}'`);
    }
  });

  const inputTypes: string[] = [...(sig.required_inputs ?? []), ...(sig.optional_inputs ?? [])];
  for (const t of inputTypes) {
    if (!OBSERVATION_TYPES.includes(t)) {
      errors.push(`${sid}: unknown input type '${t}'`);
    }
  }

  // Explanation completeness is enforced at load time (prime directive 3).
  const templateVars = new Set<string>();
  const re = /\{([a-z_][a-z0-9_]*)\}/g;
  const template: string = sig.explanation_template ?? "";
  let m: RegExpExecArray | null;
  while ((m = re.exec(template)) !== null) {
    templateVars.add(m[1]);
  }
  const bindings: Record<string, any> = sig.template_bindings ?? {};
  for (const v of Array.from(templateVars).sort()) {
    if (!(v in bindings)) {
      errors.push(`${sid}: template variable '{${v}}' has no binding`);
    }
  }
  for (const v of Object.keys(bindings).sort()) {
    const b = bindings[v];
    if (!idsSeen.has(b.condition_id)) {
      errors.push(`${sid}: binding '${v}' references unknown condition id '${b.condition_id}'`);
    }
  }
}

function relBaseName(rel: string): string {
  const idx = rel.lastIndexOf("/");
  return idx === -1 ? rel : rel.slice(idx + 1);
}

/** Load and fully validate a content package via the injected reader. */
export function loadContentFromReader(reader: ContentReader): ContentHandle {
  const errors: string[] = [];
  const flags: string[] = [];

  const manifestText = reader.readPackageText("manifest.json");
  if (manifestText === null) {
    throw new ContentError(["manifest.json missing from package"]);
  }
  const manifest = JSON.parse(manifestText);
  const manifestSchema = JSON.parse(reader.readSchemaText("manifest.schema.json"));
  for (const e of validate(manifest, manifestSchema)) {
    errors.push(`manifest.json ${e}`);
  }
  if (errors.length > 0) {
    throw new ContentError(errors);
  }

  if (semverGt(semverTuple(manifest.min_engine_version), semverTuple(ENGINE_VERSION))) {
    throw new ContentError([
      `content requires engine >= ${manifest.min_engine_version}, this is ${ENGINE_VERSION}`,
    ]);
  }
  if (manifest.signing.algorithm === "none") {
    flags.push("content_unsigned");
  }

  // Hash verification: every listed file must exist and match; every JSON
  // file in the package must be listed (no smuggled content).
  const listed: Record<string, string> = {};
  for (const f of manifest.files) {
    listed[f.path] = f.sha256;
  }
  const fileTexts: Record<string, string> = {};
  for (const rel of Object.keys(listed).sort()) {
    const expected = listed[rel];
    const text = reader.readPackageText(rel);
    if (text === null) {
      errors.push(`manifest lists missing file: ${rel}`);
      continue;
    }
    fileTexts[rel] = text;
    const actual = sha256Hex(text);
    if (actual !== expected) {
      errors.push(`hash mismatch for ${rel}: manifest=${expected} actual=${actual}`);
    }
  }
  const onDisk: string[] = [];
  for (const rel of reader.listPackageFiles()) {
    const name = relBaseName(rel);
    if (name.endsWith(".json") && name !== "manifest.json") {
      onDisk.push(rel);
    }
  }
  for (const rel of [...onDisk].sort()) {
    if (!(rel in listed)) {
      errors.push(`file present but not in manifest: ${rel}`);
    }
  }
  const fileList = Object.keys(listed)
    .map((p) => ({ path: p, sha256: listed[p] }))
    .sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
  const expectedPkgHash = sha256Hex(canonicalJson(fileList));
  if (expectedPkgHash !== manifest.package_hash) {
    errors.push("package_hash does not match file list");
  }
  if (errors.length > 0) {
    throw new ContentError(errors);
  }

  const tablesSchema = JSON.parse(reader.readSchemaText("content_tables.schema.json"));
  const signatureSchema = JSON.parse(reader.readSchemaText("signature.schema.json"));

  const tables: Record<string, any> = {};
  const signatures: any[] = [];
  for (const rel of Object.keys(listed).sort()) {
    const doc = JSON.parse(fileTexts[rel]);
    if (rel.startsWith("signatures/")) {
      const errs = validate(doc, signatureSchema);
      for (const e of errs) {
        errors.push(`${rel} ${e}`);
      }
      if (errs.length === 0) {
        signatures.push(doc);
      }
    } else {
      const ct = doc?.content_type;
      if (typeof ct !== "string" || !TABLE_TYPES.includes(ct)) {
        errors.push(`${rel}: unknown or missing content_type '${ct === undefined || ct === null ? "None" : ct}'`);
        continue;
      }
      const sub: Record<string, Json> = { $defs: tablesSchema["$defs"], $ref: `#/$defs/${ct}` };
      for (const e of validate(doc, sub, tablesSchema)) {
        errors.push(`${rel} ${e}`);
      }
      if (ct in tables) {
        errors.push(`duplicate content_type '${ct}' (${rel})`);
      }
      tables[ct] = doc;
    }
  }
  for (const ct of TABLE_TYPES) {
    if (!(ct in tables)) {
      errors.push(`package is missing required table content_type '${ct}'`);
    }
  }
  if (errors.length > 0) {
    throw new ContentError(errors);
  }

  const bounds = tables["operational_bounds"];
  const knownMetrics = new Set<string>([...Object.keys(bounds.bounds), ...Object.keys(bounds.enums)]);
  const enumMetrics = new Set<string>(Object.keys(bounds.enums));
  const eventIds = new Set<string>(tables["event_codes"].codes.map((c: any) => c.event_id));

  const seenIds = new Set<string>();
  for (const sig of signatures) {
    if (seenIds.has(sig.signature_id)) {
      errors.push(`duplicate signature_id '${sig.signature_id}'`);
    }
    seenIds.add(sig.signature_id);
    checkSignatureSemantics(sig, knownMetrics, eventIds, enumMetrics, errors);
  }
  if (errors.length > 0) {
    throw new ContentError(errors);
  }

  signatures.sort((a, b) => (a.signature_id < b.signature_id ? -1 : a.signature_id > b.signature_id ? 1 : 0));
  return {
    package_path: reader.packagePath,
    content_version: manifest.package_version,
    flags,
    tables,
    signatures,
    published_signatures: signatures.filter((s) => s.status === "published"),
  };
}

File: sentinel/engine-py/sentinel/content.py
"""Content package loading: schema validation, manifest hash verification,
semantic cross-checks. Fails closed — any problem raises ContentError with
every message collected (prime directives 2 and 4).
"""
from __future__ import annotations

import json
import os

from .canonical import canonical_json, sha256_hex
from .constants import ENGINE_VERSION, NUMERIC_TYPES, OBSERVATION_TYPES, REFERENCEABLE_FLAGS
from .errors import ContentError
from .schema_validator import validate

TABLE_TYPES = [
    "operational_bounds", "never_ignore", "floors", "baseline_config",
    "population_baselines", "units", "worse_direction", "artifact_rules",
    "global_modifiers", "recheck_intervals", "trajectory_rules", "event_codes",
    "cadence", "stream_screening",
]

COMPARATOR_OPS = ["gte", "lte", "eq"]
FUNCTION_OPS = [
    "delta_from_baseline_abs", "delta_from_baseline_pct", "trend_slope",
    "crossed_threshold_count", "sustained_for_min", "event_present",
]


def _reject_constant(name):
    # Python's json accepts NaN/Infinity by default; engine-ts (JSON.parse)
    # never can, so such content must fail closed here too.
    raise ValueError(f"non-finite number ({name}) is not valid content JSON")


def _read_json(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f, parse_constant=_reject_constant)


def _semver_tuple(v: str):
    return tuple(int(x) for x in v.split("."))


def _walk_conditions(node, out):
    """Flatten a logic tree into its leaf conditions (depth-first, in order)."""
    if not isinstance(node, dict):
        return
    for comb in ("all_of", "any_of", "none_of"):
        if comb in node:
            for child in node[comb]:
                _walk_conditions(child, out)
            return
    if "at_least_n_of" in node:
        for child in node["at_least_n_of"]["of"]:
            _walk_conditions(child, out)
        return
    out.append(node)


_COMBINATORS = ("all_of", "any_of", "none_of", "at_least_n_of")


def _check_logic_structure(node, sid, errors):
    """Every node must be a leaf or carry EXACTLY one combinator — a second
    combinator key would otherwise be silently dropped at evaluation time."""
    if not isinstance(node, dict):
        return
    present = [k for k in _COMBINATORS if k in node]
    if len(present) > 1:
        errors.append(f"{sid}: logic node has multiple combinator keys ({','.join(present)})")
        return
    if not present:
        return
    comb = present[0]
    if comb == "at_least_n_of":
        spec = node[comb]
        if isinstance(spec, dict) and isinstance(spec.get("of"), list):
            if isinstance(spec.get("n"), int) and spec["n"] > len(spec["of"]):
                errors.append(f"{sid}: at_least_n_of n={spec['n']} exceeds available conditions ({len(spec['of'])})")
            for child in spec["of"]:
                _check_logic_structure(child, sid, errors)
        return
    if isinstance(node[comb], list):
        for child in node[comb]:
            _check_logic_structure(child, sid, errors)


def _check_signature_semantics(sig, known_metrics, event_ids, enums, errors):
    sid = sig.get("signature_id", "?")
    enum_metrics = set(enums)
    _check_logic_structure(sig.get("logic", {}), sid, errors)
    conditions = []
    _walk_conditions(sig.get("logic", {}), conditions)
    if not conditions:
        errors.append(f"{sid}: logic tree has no conditions")
    uses_event_present = False
    ids_seen = set()
    for i, cond in enumerate(conditions):
        where = f"{sid}: condition {i}"
        if "flag" in cond:
            if cond["flag"] not in REFERENCEABLE_FLAGS:
                errors.append(f"{where}: unknown flag '{cond['flag']}'")
            continue
        if "baseline_status" in cond and "op" in cond:
            # The op path would win and the guard would be silently discarded.
            errors.append(f"{where}: condition must not combine 'baseline_status' with 'op'")
            continue
        if "baseline_status" in cond:
            if "metric" not in cond:
                errors.append(f"{where}: baseline_status condition requires 'metric'")
            elif cond["metric"] not in known_metrics:
                errors.append(f"{where}: unknown metric '{cond['metric']}'")
            continue
        op = cond.get("op")
        metric = cond.get("metric")
        if op is None:
            errors.append(f"{where}: leaf condition needs 'op', 'flag', or 'baseline_status'")
            continue
        cid = cond.get("id")
        if cid:
            if cid in ids_seen:
                errors.append(f"{sid}: duplicate condition id '{cid}'")
            ids_seen.add(cid)
        if op == "event_present":
            uses_event_present = True
            if "event_id" not in cond:
                errors.append(f"{where}: event_present requires 'event_id'")
            elif cond["event_id"] not in event_ids:
                errors.append(f"{where}: event_id '{cond['event_id']}' not in controlled vocabulary")
            if "window_min" not in cond:
                errors.append(f"{where}: event_present requires 'window_min'")
            continue
        if metric is None:
            errors.append(f"{where}: op '{op}' requires 'metric'")
            continue
        if metric not in known_metrics:
            errors.append(f"{where}: unknown metric '{metric}'")
        comparators = [k for k in COMPARATOR_OPS if k in cond]
        if op in COMPARATOR_OPS:
            if "value" not in cond:
                errors.append(f"{where}: raw-value op '{op}' requires 'value'")
            elif metric in enum_metrics:
                if cond["value"] not in enums.get(metric, []):
                    errors.append(f"{where}: value is not a valid level for ordinal metric '{metric}'")
            elif isinstance(cond["value"], bool) or not isinstance(cond["value"], (int, float)):
                # A string value would compare against numbers at runtime:
                # TypeError in Python, silent nonsense in JS.
                errors.append(f"{where}: value must be numeric for metric '{metric}'")
            if comparators:
                errors.append(f"{where}: raw-value op '{op}' must not also carry comparator keys")
            if "window_min" not in cond:
                errors.append(f"{where}: raw-value op '{op}' requires 'window_min'")
        elif op in FUNCTION_OPS:
            if len(comparators) != 1:
                errors.append(f"{where}: op '{op}' requires exactly one of gte/lte/eq keys")
            if "window_min" not in cond:
                errors.append(f"{where}: op '{op}' requires 'window_min'")
            if op == "trend_slope" and "min_points" not in cond:
                errors.append(f"{where}: trend_slope requires 'min_points'")
            if op in ("crossed_threshold_count", "sustained_for_min"):
                if "threshold" not in cond or "direction" not in cond:
                    errors.append(f"{where}: op '{op}' requires 'threshold' and 'direction'")
            if metric in enum_metrics and op != "event_present" and op not in COMPARATOR_OPS:
                errors.append(f"{where}: op '{op}' not valid for ordinal metric '{metric}'")
        else:
            errors.append(f"{where}: unknown op '{op}'")

    if uses_event_present and "event" not in sig.get("required_inputs", []):
        # Without this, an absent event feed silently reads as "no event"
        # instead of routing through the missing-input honest-failure path.
        errors.append(f"{sid}: logic uses event_present but required_inputs does not list 'event'")

    for t in sig.get("required_inputs", []) + sig.get("optional_inputs", []):
        if t not in OBSERVATION_TYPES:
            errors.append(f"{sid}: unknown input type '{t}'")

    # Explanation completeness is enforced at load time (prime directive 3).
    import re as _re
    template_vars = set(_re.findall(r"\{([a-z_][a-z0-9_]*)\}", sig.get("explanation_template", "")))
    bindings = sig.get("template_bindings", {})
    for var in sorted(template_vars):
        if var not in bindings:
            errors.append(f"{sid}: template variable '{{{var}}}' has no binding")
    for var, b in sorted(bindings.items()):
        if b["condition_id"] not in ids_seen:
            errors.append(f"{sid}: binding '{var}' references unknown condition id '{b['condition_id']}'")


def load_content(package_path: str, schema_dir: str | None = None) -> dict:
    """Load and fully validate a content package. Returns a ContentHandle dict."""
    if schema_dir is None:
        schema_dir = os.path.normpath(os.path.join(package_path, "..", "..", "schema"))
    errors: list[str] = []
    flags: list[str] = []

    manifest_path = os.path.join(package_path, "manifest.json")
    if not os.path.exists(manifest_path):
        raise ContentError(["manifest.json missing from package"])
    manifest = _read_json(manifest_path)
    manifest_schema = _read_json(os.path.join(schema_dir, "manifest.schema.json"))
    errors.extend(f"manifest.json {e}" for e in validate(manifest, manifest_schema))
    if errors:
        raise ContentError(errors)

    if _semver_tuple(manifest["min_engine_version"]) > _semver_tuple(ENGINE_VERSION):
        raise ContentError([
            f"content requires engine >= {manifest['min_engine_version']}, this is {ENGINE_VERSION}"
        ])
    if manifest["signing"]["algorithm"] == "none":
        flags.append("content_unsigned")

    # Hash verification: every listed file must exist and match; every JSON
    # file in the package must be listed (no smuggled content).
    listed = {f["path"]: f["sha256"] for f in manifest["files"]}
    for rel, expected in sorted(listed.items()):
        full = os.path.join(package_path, rel)
        if not os.path.exists(full):
            errors.append(f"manifest lists missing file: {rel}")
            continue
        with open(full, "rb") as fh:
            actual = sha256_hex(fh.read().decode("utf-8"))
        if actual != expected:
            errors.append(f"hash mismatch for {rel}: manifest={expected} actual={actual}")
    on_disk = []
    for root, _dirs, files in os.walk(package_path):
        for name in files:
            if name.endswith(".json") and name != "manifest.json":
                on_disk.append(os.path.relpath(os.path.join(root, name), package_path).replace(os.sep, "/"))
    for rel in sorted(on_disk):
        if rel not in listed:
            errors.append(f"file present but not in manifest: {rel}")
    expected_pkg_hash = sha256_hex(canonical_json(sorted(
        [{"path": p, "sha256": h} for p, h in listed.items()], key=lambda x: x["path"]
    )))
    if expected_pkg_hash != manifest["package_hash"]:
        errors.append("package_hash does not match file list")
    if errors:
        raise ContentError(errors)

    tables_schema = _read_json(os.path.join(schema_dir, "content_tables.schema.json"))
    signature_schema = _read_json(os.path.join(schema_dir, "signature.schema.json"))

    tables: dict = {}
    signatures: list = []
    for rel in sorted(listed):
        doc = _read_json(os.path.join(package_path, rel))
        if rel.startswith("signatures/"):
            errs = validate(doc, signature_schema)
            errors.extend(f"{rel} {e}" for e in errs)
            if not errs:
                signatures.append(doc)
        else:
            ct = doc.get("content_type")
            if ct not in TABLE_TYPES:
                errors.append(f"{rel}: unknown or missing content_type '{ct}'")
                continue
            sub = {"$defs": tables_schema["$defs"], "$ref": f"#/$defs/{ct}"}
            errors.extend(f"{rel} {e}" for e in validate(doc, sub, tables_schema))
            if ct in tables:
                errors.append(f"duplicate content_type '{ct}' ({rel})")
            tables[ct] = doc
    for ct in TABLE_TYPES:
        if ct not in tables:
            errors.append(f"package is missing required table content_type '{ct}'")
    if errors:
        raise ContentError(errors)

    bounds = tables["operational_bounds"]
    known_metrics = set(bounds["bounds"].keys()) | set(bounds["enums"].keys())
    event_ids = {c["event_id"] for c in tables["event_codes"]["codes"]}

    seen_ids = set()
    for sig in signatures:
        if sig["signature_id"] in seen_ids:
            errors.append(f"duplicate signature_id '{sig['signature_id']}'")
        seen_ids.add(sig["signature_id"])
        _check_signature_semantics(sig, known_metrics, event_ids, bounds["enums"], errors)
    if errors:
        raise ContentError(errors)

    signatures.sort(key=lambda s: s["signature_id"])
    return {
        "package_path": package_path,
        "content_version": manifest["package_version"],
        "flags": flags,
        "tables": tables,
        "signatures": signatures,
        "published_signatures": [s for s in signatures if s["status"] == "published"],
    }
